File: CombSum.java
class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result= new ArrayList<>();
        if (candidates.length==0 || candidates==null) return result;
        
        helper(candidates, 0, target, new ArrayList<>());
        return result;
    }
    
    private void helper(int[] candidates, int pivot, int amount, List<Integer> path){
        
        //base case
        if(amount ==0){
            result.add(path);
            return;
        }
        if(amount<0 || pivot==candidates.length){
            return;
        }
        
        //logic
        for(int i=pivot; i<candidates.length;i++){
            path.add(candidates[i]);
            helper(candidates, i, amount-candidates[i], new ArrayList<>(path));
            path.remove(path.size()-1);
        }
        
    }
    
}

File: combSum.py
#Time complexity: O(2^(Target/minNum))
#space complexity: O(Target/minNum)
class Solution:
    def __init__(self):
        self.res = []
        
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        self.helper(candidates, target, 0, [])
        return self.res
    
    def helper(self, candidates, target, idx, path):
        #base
        if target == 0:
            self.res.append(path.copy())
            return
        if target < 0:
            return
        
        #logic
        for i in range(idx, len(candidates)):
            path.append(candidates[i])
            self.helper(candidates, target-candidates[i], i, path)
            del path[-1]
            
        
        
